### PR TITLE
warn if image is not a builder, even if it came from source detection

### PR DESCRIPTION
**Pull request by bparees**
fixes https://github.com/openshift/origin/issues/4207

### CODE REVIEW

**Comment by bparees** [comment]
@csrwng ptal, this should help with the cases where someone doesn't have any imagestreams defined and we match to the dockerhub "ruby" image as an s2i build, and then the build fails horribly.

**Comment by bparees** [comment]
sample:
```
$ oc new-app https://github.com/openshift/rails-ex --dry-run
error: the resolved match "ruby" for component "ruby" cannot build source code - check whether this is the image you want to use, then use --strategy=source to build using source or --strategy=docker to treat this as a Docker base image and set up a layered Docker build
```

then run with --strategy=source (which will result in a failing s2i build):

```
sample-app (source_warn)$ oc new-app https://github.com/openshift/rails-ex --dry-run --strategy=source
--> Found Docker image 7a617c7 (7 days old) from Docker Hub for "ruby"

    * An image stream will be created as "ruby:latest" that will track the source image
    * The source repository appears to match: ruby
    * A source build using source code from https://github.com/openshift/rails-ex will be created
      * The resulting image will be pushed to image stream "rails-ex:latest"
      * Every time "ruby:latest" changes a new build will be triggered
    * This image will be deployed in deployment config "rails-ex"
    * The image does not expose any ports - if you want to load balance or send traffic to this component
      you will need to create a service with 'expose dc/rails-ex --port=[port]' later
    * WARNING: Image "rails-ex" runs as the 'root' user which may not be permitted by your cluster administrator

--> Creating resources with label app=rails-ex ...
--> Success (DRY RUN)
```

**Comment by bparees** [comment]
[test]

File: pkg/generate/app/cmd/newapp.go [hunk 1]
```diff
@@ -243,6 +243,34 @@ func (c *AppConfig) AddArguments(args []string) []string {
...
+// validateBuilders confirms that all images associated with components that are to be built,
+// are builders (or we're using a docker strategy).
+func (c *AppConfig) validateBuilders(components app.ComponentReferences) error {
+	errs := []error{}
+	for _, ref := range components {
+		input := ref.Input()
+
+		// if the strategy is explicitly Docker, all repos should assume docker
+		if c.Strategy == "docker" && input.Uses != nil {
+			input.Uses.BuildWithDocker()
```
Comment (bparees) [inline pkg/generate/app/cmd/newapp.go]: yeah, this code had rotted away in a forgotten branch for me, so it turns out that logic was redundant now.  will clean it up, thanks for catching it.

File: pkg/generate/app/cmd/newapp.go [hunk 2]
```diff
@@ -243,6 +243,34 @@ func (c *AppConfig) AddArguments(args []string) []string {
...
+		}
+
+		// if we are expecting build inputs, or get a build input when strategy is not docker, expect to build
+		if c.ExpectToBuild || (input.ResolvedMatch.Builder && c.Strategy != "docker") {
+			input.ExpectToBuild = true
+		}
+
+		switch {
+		case input.ExpectToBuild && input.ResolvedMatch != nil && !input.ResolvedMatch.Builder && input.Uses != nil && !input.Uses.IsDockerBuild():
+			if len(c.Strategy) == 0 {
```
Comment (bparees) [inline pkg/generate/app/cmd/newapp.go]: yeah, i think i had a greater vision once upon a time for what this would check :)

**Comment by bparees** [comment]
@csrwng hopefully better now.

**Comment by csrwng** [comment]
Looks like the TestNewAppNewBuildEnvVars integration test needs to be fixed. Other than that, LGTM

**Comment by mfojtik** [comment]
@bparees does this mean I need to set label for my docker image to tell OpenShift it is a builder? (maybe i am just not familiar with how the IsBuilder works).

**Comment by bparees** [comment]
@mfojtik you need a tag on your imagestream:
https://github.com/openshift/origin/blob/master/examples/image-streams/image-streams-centos7.json#L20

Note that you also need the "ruby" tag for us to have even matched your source repo to that image.

I suppose this could cause confusion if someone creates their own imagestream and then directly attempts to use it (`oc new-app myimagestreamtag~myrepo`) at which point they'll get this message.  I'll take another spin at this and see if i can distinguish between "stuff we found" and "stuff they explicitly asked for".

it's not going into 3.2 regardless.

**Comment by bparees** [comment]
> I suppose this could cause confusion if someone creates their own imagestream and then directly attempts to use it (oc new-app myimagestreamtag~myrepo) at which point they'll get this message. I'll take another spin at this and see if i can distinguish between "stuff we found" and "stuff they explicitly asked for".

so the new-app logic already checks for two things to determine if an image is a builder:
1) the builder tag on the imagestream
2) the image itself looks like a builder based on the s2i scripts label

So given that, i think this is good as is.  They won't get the warning if they create their own imagestream and it gets chosen (either explicitly or implicitly), as long as the image contains the s2i label (per (2) above).  If it doesn't contain that label, it's probably appropriate to warn them, again, whether or not the explicitly chose the image.

**Comment by smarterclayton** [comment]
That's a pretty long error message.  If it's important to show it, it should be better formatted for the end user

**Comment by bparees** [comment]
it is important, the scenario here is that the user is attempting to use a non-s2i builder with their source code.  So there is a lot of information to convey:

1) which image new-app was trying to use and for which input/component
2) the fact that this image is not a source builder
3) how to override new-app and use it as a source builder anyway
4) how to do what is probably the right thing and use it for a docker builder instead.

So yes it's going to be a long message.  Aside from quoting the 2 argument strings, i don't know what else you'd like formatted though.  It would also be nice if instead of "component ruby" it said "for repo foo.git" i suppose.

**Comment by bparees** [comment]
@smarterclayton how about this:
```
$ oc new-app https://github.com/openshift/rails-ex --dry-run
error: the image match "ruby" for source repository "https://github.com/openshift/rails-ex" cannot build source code - check if this is the image you want to use, then pass "--strategy=source" to use it as a source builder image or "--strategy=docker" to use it as a base image for a Docker build
```

**Comment by smarterclayton** [comment]
I feel like this needs a formatted message, not just the really long
multi-line wrapping

**Comment by bparees** [comment]
@smarterclayton 
```
$ oc new-app https://github.com/openshift/rails-ex --dry-run
error: the image match "ruby" for source repository "https://github.com/openshift/rails-ex" does not appear to be a source-to-image builder.

- to attempt to use this image as a source builder, pass "--strategy=source"
- to use it as a base image for a Docker build, pass "--strategy=docker"
```

**Comment by bparees** [comment]
@smarterclayton bump for review of the new warning message (see above).

**Comment by bparees** [comment]
@smarterclayton https://github.com/openshift/origin/pull/8323#issuecomment-208660113

**Comment by bparees** [comment]
[merge]

**Pull request merged**